### PR TITLE
Refactor EkuboLiquidity for better readability and maintenance

### PR DESCRIPTION
**Pull request by vitaliypopel**
**Whats was did?**
- Was created EkuboLiquidity class in the utils.py file. 
- Implemented the Constructor: created initializer method __init__ in the class with the required parameters: data, collateral_token, and debt_token. And additional params for different fetches, params_for_bids and params_for_asks.
- Created method apply_liquidity_to_dataframe which taking liquidity data (bids or asks) and process it into a Pandas dataframe for further analysis or usage.
- Created fetch_liquidity method which implements fetch liquidity data from the given URL and returning structured dictionary for work with.
- Migrated _get_available_liquidity which was copied from app.py origionally.
- Also migrated _remove_leading_zeros as staticmethod for easy work.

Note: __All works good!__

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                price_diff=price_diff,
+                bids=True if bids_or_asks['type'] == 'bids' else False,
+            ),
+        )
+        self.data['debt_token_supply'] += self.data['Ekubo_debt_token_supply']
+
+        return self.data
+
+    def fetch_liquidity(self, bids: bool = True) -> dict[str, Any]:
+        params = self.params_for_bids
```
Comment (djeck1432) [inline utils.py:62]: add docstring to this method

File: utils.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        if not bids:
+            params = self.params_for_asks
+            logging.warning('Using collateral token as base token and debt token as quote token.')
+
+        response = requests.get(self.URL, params=params)
+
+        if response.status_code == 200:
```
Comment (djeck1432) [inline utils.py:69]:
```suggestion
        if response.ok:
```

File: utils.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        else:
+            self.fetch_liquidity()
+
+    @staticmethod
+    def _get_available_liquidity(
+            data: pandas.DataFrame,
+            price: float,
+            price_diff: float,
+            bids: bool
+    ) -> float:
```
Comment (djeck1432) [inline utils.py:92]: add docstring

File: utils.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        price_lower_bound = max(0.95 * price, price - price_diff) if bids else price
```
Comment (djeck1432) [inline utils.py:93]: move to class variable this value `0.95`

File: utils.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        price_upper_bound = price if bids else min(1.05 * price, price + price_diff)
```
Comment (djeck1432) [inline utils.py:94]: move to class variable this value `1.05`

File: utils.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        return data.loc[
+            data['price'].between(
+                price_lower_bound,
+                price_upper_bound,
+            ),
+            'quantity',
+        ].sum()
+
+    @staticmethod
+    def _remove_leading_zeros(address: str) -> str:
```
Comment (djeck1432) [inline utils.py:104]: add docstring

**Review by djeck1432 — CHANGES_REQUESTED** [review]

**Review by djeck1432 — APPROVED** [review]
Looks good to me

File: utils.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        if response.ok:
+            liquidity = response.json()
+            data = {
+                "type": "bids" if bids else "asks",
+            }
+            try:
+                data["prices"], data["quantities"] = zip(*liquidity[data["type"]])
+            except ValueError:
+                time.sleep(300 if bids else 5)
+                self.fetch_liquidity()
```
Comment (lukaspetrasek) [inline utils.py:104]: Pass in the argument, too.

File: app.py [hunk 8]
```diff
@@ -196,14 +106,20 @@ def process_liquidity(main_chart_data: pandas.DataFrame, collateral_token: str,
...
-    main_chart_data = add_ekubo_liquidity(
+
+    ekubo_liquidity = EkuboLiquidity(
         data=main_chart_data,
         collateral_token=collateral_token_underlying_address,
         debt_token=debt_token_underlying_address,
+
+    )
+    main_chart_data = ekubo_liquidity.apply_liquidity_to_dataframe(
+        ekubo_liquidity.fetch_liquidity(),
```
Comment (lukaspetrasek) [inline app.py:117]: Wait, so this is now fetching only bids? Why not asks?

File: utils.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+            else:
+                return data
+        elif bids:
+            self.fetch_liquidity(bids=False)
+        else:
+            self.fetch_liquidity()
```
Comment (lukaspetrasek) [inline utils.py:110]: Maybe add some small sleep here?

File: utils.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    def apply_liquidity_to_dataframe(
+        self, bids_or_asks: dict[str, Any]
+    ) -> pandas.DataFrame:
+        """
+        Applying liquidity bids or asks data to dataframe, saving in object and returns data
+        :param bids_or_asks: dict[str, Any]
+        :return: pandas.DataFrame
+        """
+
+        liquidity_dataframe = pandas.DataFrame(
```
Comment (lukaspetrasek) [inline utils.py:49]:
```suggestion
        liquidity = pandas.DataFrame(
```
.. No need to put the type into the name 😉

File: app.py [hunk 11]
```diff
@@ -406,54 +311,43 @@ def _get_risk_level(debt_to_supply_ratio: float) -> str:
...
         if not protocol:
             streamlit.write(f'Selected random protocol = {random_protocol}.')
             protocol = random_protocol
-
     loan = loans_data.loc[
         (loans_data['User'] == user)
         & (loans_data['Protocol'] == protocol),
     ]
-
-    if loan.empty:
```
[the remark this answers is not in the document]
Comment (djeck1432) [reply app.py:431]: It seems to me that it's merge conflict fix is bad, I noticed that this line was removed in another guy as well. 
fyi @lukaspetrasek

File: app.py [hunk 12]
```diff
@@ -347,24 +263,13 @@ def _get_risk_level(debt_to_supply_ratio: float) -> str:
...
             f"lending protocols is {_get_risk_level(example_row['debt_to_supply_ratio'])}."
-        )    
+        )
         streamlit.write(
             f"The ratio of liquidated debt to available supply is {round(example_row['debt_to_supply_ratio'] * 100)}%.Debt"
             f" worth of {int(example_row['liquidable_debt_at_interval']):,} USD will be liquidated while the AMM swaps "
             f"capacity will be {int(example_row['debt_token_supply']):,} USD."
         )
 
-    streamlit.header("Liquidable debt")
```
[the remark this answers is not in the document]
Comment (vitaliypopel) [reply app.py:357]: returned

File: app.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (vitaliypopel) [reply app.py]: returned

File: app.py (hunk 8 above)
Comment (vitaliypopel) [reply app.py:116]: As it was in add_ekubo_liquidity it's automatically getting or bids or asks.

**Pull request merged**